tokio/src/net/tcp/listener.rs

```diff
@@ -203,13 +203,54 @@ impl TcpListener {
     ///
     /// # Notes
     ///
-    /// The caller is responsible for ensuring that the listener is in
-    /// non-blocking mode. Otherwise all I/O operations on the listener
+    /// This sets the socket to non-blocking mode if not already done, which
+    /// is necessary for normal operation within Tokio.
+    ///
+    /// # Examples
+    ///
+    /// ```rust,no_run
+    /// use std::error::Error;
+    /// use tokio::net::TcpListener;
+    ///
+    /// #[tokio::main]
+    /// async fn main() -> Result<(), Box<dyn Error>> {
+    ///     let std_listener = std::net::TcpListener::bind("127.0.0.1:0")?;
+    ///     let listener = TcpListener::from_tcp(std_listener)?;
+    ///     Ok(())
+    /// }
+    /// ```
+    ///
+    /// # Panics
+    ///
+    /// This function panics if it is not called from within a runtime with
+    /// IO enabled.
+    ///
+    /// The runtime is usually set implicitly when this function is called
+    /// from a future driven by a tokio runtime, otherwise runtime can be set
+    /// explicitly with [`Runtime::enter`](crate::runtime::Runtime::enter) function.
+    #[track_caller]
+    pub fn from_tcp(listener: net::TcpListener) -> io::Result<TcpListener> {
+        listener.set_nonblocking(true)?;
+        Self::from_tcp_unchecked(listener)
+    }
+
+    /// Creates new `TcpListener` from a `std::net::TcpListener` without
+    /// checking that it's non-blocking.
+    ///
+    /// This function is intended to be used to wrap a TCP listener from the
+    /// standard library in the Tokio equivalent. However, it does not check
+    /// that it's already in non-blocking mode.
+    ///
+    /// Instead, the caller is responsible for ensuring that the listener is in
+    /// non-blocking mode, otherwise all I/O operations on the listener
     /// will block the thread, which will cause unexpected behavior.
     /// Non-blocking mode can be set using [`set_nonblocking`].
     ///
     /// [`set_nonblocking`]: std::net::TcpListener::set_nonblocking
     ///
+    /// You may generally prefer using [`from_tcp`](TcpListener::from_tcp),
+    /// which does that for you if not already done.
```

Comment:
```suggestion
    /// You should usually prefer to use [`from_tcp`](TcpListener::from_tcp).
```

Reply:
Small nit: for reference documentation, I think it's better to avoid "you". I'd rewrite this as "It may be preferable to use [`from_tcp`](TcpListener::from_tcp), which sets `set_nonblocking`."